src/onnx/parse_scan.cpp

```diff
@@ -0,0 +1,314 @@
+                                    size_t rank_offset) const
+    {
+        auto axes = parse_vector_attribute(info, name, expected_size);
+        if(axes.empty())
+            return {expected_size, 0};
+
+        std::transform(axes.begin(),
+                       axes.end(),
+                       ins_begin,
+                       axes.begin(),
+                       [&](int64_t axis, instruction_ref arg) {
+                           return normalize_axis(axis, arg->get_shape().ndim() + rank_offset, name);
+                       });
+
+        return axes;
+    }
+
+    // Alter the Scan body to match a body that Loop would expect.
+    //
+    // Loop body inputs: iteration_num, condition, loop_state_variables
+    // Scan body inputs: loop_state_variables, scan_input_slices
+    // iteration_num and condition parameters are prepended to the Scan body parameter list, while
+    // scan_input_slices are removed from parameters.
+    // Instead, scan_inputs are used directly in Scan body(as values from enclosing scope), and
+    // together with iteration_num passed to the scan_slice operator which produces slices that are
+    // used instead of the scan_inputs_slices.
+    //
+    // Loop body outputs: condition, loop_state_variables, scan_output_slices
+    // Scan body outputs: loop_state_variables, scan_output_slices
+    // The inserted Scan body condition parameter is prepended to the Scan body returns
+    void modify_body(module_ref mod,
+                     const std::vector<instruction_ref>& args,
+                     int64_t n,
+                     int64_t m,
+                     const std::vector<int64_t>& scan_input_axes,
+                     const std::vector<int64_t>& scan_input_directions) const
+    {
+        std::vector<instruction_ref> params;
+        params.reserve(n + m);
+        transform(mod->get_parameter_names(),
+                  std::back_inserter(params),
+                  [&](const std::string& name) { return mod->get_parameter(name); });
+
+        // iteration_num, condition, and duplicate loop_state_variables are appended to parameters.
+        // References to the original loop_state_variables in other instructions are then replaced
+        // with references to the duplicate ones, after which the originals are removed.
+        //
+        // References to the scan_input_slices are replaced with references to inserted
+        // scan_slice->squeeze instructions, after which the scan_input_slices parameters are
+        // removed.
+        auto iter_param = mod->add_parameter("iter", shape{shape::int64_type});
+        auto cond_param = mod->add_parameter("cond", shape{shape::bool_type});
+        std::vector<instruction_ref> new_params;
+        new_params.reserve(n);
+        for(auto i = 0; i < n; ++i)
+            new_params.push_back(
+                mod->add_parameter("state_var" + std::to_string(i), params[i]->get_shape()));
```

Comment:
you can use `std::transform` here

Reply:
I use the loop iterator value in the loop body within the std::to_string as well, so this isn't that well suited to std::transform.